test/unit/Cardano/Wallet/Api/TypesSpec.hs

```diff
@@ -164,3 +208,119 @@ instance Arbitrary WalletState where
 instance Arbitrary a => Arbitrary (ApiT a) where
     arbitrary = ApiT <$> arbitrary
     shrink = fmap ApiT . shrink . getApiT
+
+
+{-------------------------------------------------------------------------------
+                   Specification / Servant-Swagger Machinery
+
+  Below is a bit of complicated API-Level stuff in order to achieve two things:
+
+  1/ Verify that every response from the API that actually has a JSON content
+     type returns a JSON instance that matches the JSON format described by the
+     specification (field names should be the same, and constraints on values as
+     well).
+     For this, we need three things:
+         - ToJSON instances on all those types, it's a given with the above
+         - Arbitrary instances on all those types, that reflect as much as
+           possible, all possible values of those types. Also given by using
+           'genericArbitrary' whenever possible.
+         - ToSchema instances which tells how do a given type should be
+           represented.
+     The trick is for the later point. In a "classic" scenario, we would have
+     defined the `ToSchema` instances directly in Haskell on our types, which
+     eventually becomes a real pain to maintain. Instead, we have written the
+     spec by hand, and we want to control that our implementation matches it.
+     So, we "emulate" the 'ToSchema' instance by:
+         - Parsing the specification file (which is embedded at compile-time)
+         - Creating missing 'ToSchema' by doing lookups in that global schema
+
+  2/ The above verification is rather week, because it just controls the return
+     types of endpoints, but not that those endpoints are somewhat valid. Thus,
+     we've also built another check 'validateEveryPath' which crawl our servant
+     API type, and check whether every path we have in our API appears in the
+     specification. It does it by defining a few recursive type-classes to
+     crawl the API, and for each endpoint:
+         - construct the corresponding path (with verb)
+         - build an HSpec scenario which checks whether the path is present
+    This seemingly means that the identifiers we use in our servant paths (in
+    particular, those for path parameters) should exactly match the specs.
```

Comment:
I guess at some point we may want to add another test, namely that:

> "every path that exists in the Swagger specification also exists as a path in our Servant API."

But such a test will **fail** _unless_ we really do have a complete API, so we can only add it when we do have  a complete API.

Reply:
Yes. Thought of it indeed but in practice, we will likely modify and extend the specification always before adding the corresponding implementation. We could also just have "stubs" in the Haskell with empty handlers throwing 501. It could work quite fine without much overhead when modifying specs.